loki/program_unit.py

```diff
@@ -54,16 +57,20 @@ class ProgramUnit(Scope):
         Mark the object as incomplete, i.e. only partially parsed. This is
         typically the case when it was instantiated using the :any:`Frontend.REGEX`
         frontend and a full parse using one of the other frontends is pending.
+    parser_classes : :any:`RegexParserClass`, optional
+        Provide the list of parser classes used during incomplete regex parsing
     """
 
     def __initialize__(self, name, docstring=None, spec=None, contains=None,
-                       ast=None, source=None, rescope_symbols=False, incomplete=False):
+                       ast=None, source=None, rescope_symbols=False, incomplete=False,
+                       parser_classes=None):
         # Common properties
         assert name and isinstance(name, str)
         self.name = name
         self._ast = ast
         self._source = source
         self._incomplete = incomplete
+        self._parser_classes = parser_classes
```

Comment:
I'm not sure I understand why the `ParserClass` (a static property, no?) needs to be passed through to the object constructor all of a sudden. Should the regex parser class not be known statically from the type? What am I missing?

Reply:
Ah, apologies, that wasn't explained very well.

This property does not represent the `ParserClass` that is responsible for matching the specific `ProgramUnit`, but is for book-keeping which parser classes were active when the code has been parsed with the REGEX frontend. The rationale is that we do now (in fact, did before but now it's tested and exposed) support an incremental parsing with the REGEX frontend. For example:

In a first round, we parse 10 files with the REGEX frontend and only using `RegexParserClass.ProgramUnitClass`. That will match the `Module` and `Subroutine` objects inside each of them and not recurse any further.
Then, we take one of these routines ("seed") and trigger a new parse of that file with additional parser classes, say `ImportClass | CallClass`. It will now match additional IR nodes inside the previously matched ProgramUnit, and update this property to include all parser classes that have been used so far. If the re-parse is requested with a subset of already used parser classes, this is a no-op and we return directly.

There's a corresponding test that illustrates the behaviour:

https://github.com/ecmwf-ifs/loki/blob/15f0de3da52d71d0c51c643f095e8f2966070281/tests/test_frontends.py#L608-L754

The new Scheduler builds upon this behaviour when populating the dependency graph. With the IR nodes matched that constitutes dependencies, we pick out the corresponding scope objects and repeat the exercise there: match new IR nodes, find their dependencies, pick out the next scope objects.